webapp/store/selectors/nav.js

```diff
@@ -0,0 +1,176 @@
+
+  const sortedPluginMetadata = pluginMeta
+    .filter(plugin => {
+      if (plugin.parent) {
+        // if it has a parent then add it to corresponding array in subItems map
+        // this is so that we can get all subItems and then sort later
+        const { parent } = plugin;
+        if (!subItems.has(parent)) {
+          subItems.set(parent, [plugin]);
+        } else {
+          const children = subItems.get(parent);
+          children.push(plugin);
+          subItems.set(parent, children);
+        }
+        // don't return to array of top level items
+        return false;
+      } else {
+        // only returning an array of the parents
+        return true;
+      }
+    })
+    .sort(comparePlugins);
+
+  subItems.forEach((children, parent) => {
+    const index = sortedPluginMetadata.findIndex(
+      plugin => parent === plugin.name
+    );
+    children.sort(comparePlugins);
+    sortedPluginMetadata.splice(index + 1, 0, ...children);
+  });
+
+  return sortedPluginMetadata;
+};
+
+/* Update list of plugin metadata (or nav objects) to have properly
+ * formatted, nested paths
+ * e.g. { name: 'child', parent: 'parent' } becomes
+ * { name: 'child', parent: 'parent', pathName: 'parent/child'}
+ * @param {Array{}} metadata - Array of plugin metadata objects
+ * @returns {Array{}} plugins - returns an array of metadata w/ updated paths
+ */
+export function getNestedPaths(metadata) {
+  assert(Array.isArray(metadata), 'Must pass array of metadata');
+  return metadata.reduce((updated, plugin) => {
+    if (plugin.parent) {
+      // if this plugin is a child then update its pathName property
+      // to nest behind the parent
+      const parentIndex = metadata.findIndex(
+        item => item.name === plugin.parent
+      );
+
+      const parent = metadata[parentIndex];
+      // eslint-disable-next-line no-console
+      console.assert(
+        parent,
+        `Parent ${plugin.parent} could not be found for child ${plugin.name}`
+      );
+      // get parent path, if none, use name
+      const parentPath = parent.pathName ? parent.pathName : parent.name;
+
+      // get plugin path, if none, use name
+      let pathName = plugin.pathName ? plugin.pathName : plugin.name;
+
+      // sanitize of any other uri constructions like other nested paths
+      pathName = pathName.slice(pathName.lastIndexOf('/') + 1);
+
+      // make nested path
+      pathName = `${parentPath}/${pathName}`;
+      plugin.pathName = pathName;
+    }
+    updated.push(plugin);
+    return updated;
+  }, []);
+}
+
+/* Get plugin metadata list w/ unique pathNames
+ * @param {Object{}} metadata - object of plugin metadata objects
+ * @returns {Array{}} plugins - returns an array of metadata objects with unique paths
+ */
+export const getUniquePaths = metadata => {
+  assert(Array.isArray(metadata), 'Must pass array of metadata');
+  const paths = new Set();
+  const names = new Set();
+
+  return metadata.reduce((acc, plugin) => {
```

Comment:
Using `reduce` functionally, I feel like there shouldn't be the globals `paths` and `names` but instead that information should be apart of the accumulator

Reply:
`paths` and `names` are basically just used to track for duplicates though. We don't want/need it as part of the information that's returned. It could do the checks that we are using the `Set`s for on the accumulator but the look ups wouldn't be constant since we're accumulating an array. Is there another way to use the accumulator space in reduce that would be useful for this?

Reply:
Note that this is opinionated based on my experience of programming in elixir and javascript style:

```
const { result } = metadata.reduce((acc, plugin) => {
  ... do work
  return acc;
}, { paths: new Set(), names: new Set(), result: [] })
return result;
```

Everything is kept in the scope that it is needed in and if you could `return { result }` that would be even better, but I don't think that is valid syntax

Reply:
Huh, that's clever! My only objection would be that it doesn't feel as readable at first glance, but that could just be because I'm more used to only accumulating on what you want at the very end. Anyway this works for me :)